### PR TITLE
=doc #18657 Doc java directives

### PR DESCRIPTION
**Pull request by patriknw**

### CODE REVIEW

File: akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        completeWithStatus(StatusCodes.OK));
+
+    path("test").route(matchListOfHosts);
+    //#host1
+
+    //#host2
+    final Route shortOnly = host(hostname -> hostname.length() < 10,
+        completeWithStatus(StatusCodes.OK));
+
+    path("test").route(shortOnly);
```
Comment (patriknw) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java]: Here is corresponding scala documentation: http://doc.akka.io/docs/akka-stream-and-http-experimental/snapshot/scala/http/routing-dsl/directives/host-directives/host.html#host

I don't know if this path("test") makes sense?

The Scala documentation shows test code: `Get() ~> Host("rest.company.com") ~> route ~> check`

Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java]: I think what we should show here is the javadsl testkit:

```java
        // test responses to potential errors
        appRoute.run(HttpRequest.GET("/")) // TODO make it be to host api.company.com
            .assertStatusCode(StatusCodes.OK);
```

etc

See here: https://github.com/akka/akka/blob/release-2.3-dev/akka-docs-dev/rst/java/code/docs/http/javadsl/server/testkit/TestkitExampleTest.java

File: akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package docs.http.javadsl.server.directives;
```
Comment (drewhk) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java]: how come this file is not in akka-samples/akka-docs-java-lambda? We don't need Java8 stuff?

Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java]: We're going ahead with java8 in docs here.

File: akka-http-core/src/main/java/akka/http/javadsl/model/headers/Host.java [hunk 3]
```diff
@@ -4,7 +4,22 @@
...
+        return akka.http.scaladsl.model.headers.Host.apply(address);
+    }
+    
+    public static Host create(String host) {
+        return akka.http.scaladsl.model.headers.Host.apply(host);
+    }
+    
+    public static Host create(String host, int port) {
+      return akka.http.scaladsl.model.headers.Host.apply(host, port);
+    }
```
Comment (ktoso) [inline akka-http-core/src/main/java/akka/http/javadsl/model/headers/Host.java]: Great catch, thanks @patriknw !

File: akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  @Test
+  public void testListOfHost() {
+    //#host1
+    final Route matchListOfHosts = host(
+        Arrays.asList("api.company.com", "rest.company.com"),
+        completeWithStatus(StatusCodes.OK));
+
+    testRoute(matchListOfHosts).run(
+        HttpRequest.GET("/").addHeader(Host.create("api.company.com")))
+        .assertStatusCode(StatusCodes.OK);
```
Comment (patriknw) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java]: This does not work, and I can't find what is wrong. To me it looks like the host is extracted from the uri and not from the header (as it should according to the documentation), but I can't see why the Scala test works.

File: akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java (hunk 2 above)
Comment (drewhk) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java]: that is great. Why can't we do the same for streams?

Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java]: I thought we are? The cookbook used java8

**Comment by patriknw** [comment]
I have added `RequestVal.host` (extractHost) and all method directives.

**Comment by patriknw** [comment]
I'll add the `RequestVal.matchAndExtractHost` also. It's the regexp variant of `host`.

File: akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/MethodDirectivesExamplesTest.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  // FIXME where is HttpRequest.PATCH ?
+  //  @Test
+  //  public void testPatch() {
+  //    //#patch
+  //    final Route route = patch(complete("This is a PATCH request."));
+  //
+  //    testRoute(route).run(HttpRequest.PATCH("/", "patch content")).assertEntity(
+  //        "This is a PATCH request.");
+  //    //#patch
+  //  }
```
Comment (patriknw) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/MethodDirectivesExamplesTest.java]: does anyone know?

Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/MethodDirectivesExamplesTest.java]: The request method itself is `akka.http.javadsl.model.HttpMethods.PATCH` etc.  @patriknw

Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/MethodDirectivesExamplesTest.java]: Seems like another catch actually, should be also made available on `akka.http.javadsl.model.HttpRequest` as the other methods for ease of testing.

// cc @sirthias @jrudolph

Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/MethodDirectivesExamplesTest.java]: i.e.

```
public static HttpRequest PATCH(String uri) {
  return create(uri).withMethod(HttpMethods.PATCH);
}
```

Can you add it please @patriknw ?

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3824/:

```
 16288 (akka-docs-dev/sphinx:generatePdf) Sphinx pdf generation failed.  See debug output for details.
 16289 Total time: 779 s, completed Oct 7, 2015 4:35:38 PM
```

**Comment by patriknw** [comment]
I have added docs for `RequestVals.matchAndExtractHost` and added the missing PATCH and OPTIONS.

**Comment by patriknw** [comment]
Regarding the sphinx error I can reproduce it locally, but I have no clue what is causing it.

File: akka-docs-dev/rst/java/http/routing-dsl/directives/host-directives/extractHost.rst [hunk 6]
```diff
@@ -0,0 +1,13 @@
+.. _-extractHost-java-:
```
Comment (ktoso) [inline akka-docs-dev/rst/java/http/routing-dsl/directives/host-directives/extractHost.rst]: We'll need to figure out where to document request values so it's consistent.. For now looks good :)

File: akka-docs-dev/rst/java/http/routing-dsl/directives/method-directives/index.rst [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+.. toctree::
+   :maxdepth: 1
+
+   delete
+   extractMethod
+   get
+   head
+   method
+   options
+   overrideMethodWithParameter
```
Comment (ktoso) [inline akka-docs-dev/rst/java/http/routing-dsl/directives/method-directives/index.rst]: Put this one last?

**Comment by ktoso** [comment]
LGTM, will merge once the sphinx fix is merged and this can be validated then

**Comment by ktoso** [comment]
PLS BUILD

**Comment by patriknw** [comment]
I'll squash first

**Comment by ktoso** [comment]
Great thanks

**Comment by patriknw** [comment]
Squashed and rebased. This PR includes:

=doc #18657 Doc host and path directives

* also including docs of RequestVals.matchAndExtractHost
  and RequestVals.matchAndExtractHost
* and add Host.create factory methods
* add missing HttpRequest PATCH and OPTIONS
* change to val in matchAndExtractHost for
  fail fast exception

**Comment by patriknw** [comment]
Refs #18657

**Pull request merged**